runtime/connectors/blob/blobdownloader.go

```diff
@@ -3,119 +3,272 @@ package blob
+// clients should call close once done to release all resources like closing the bucket
+// the iterator takes responsibility of closing the bucket
+func NewIterator(ctx context.Context, bucket *blob.Bucket, opts Options, globPattern string) (connectors.FileIterator, error) {
+	opts.validate()
 
-	handler := &BlobHandler{
-		prefix: prefix,
-		bucket: bucket,
-		path:   bucketPath,
+	items, err := plan(ctx, bucket, opts, globPattern)
+	if err != nil {
+		bucket.Close()
+		return nil, err
 	}
 
-	if !fileutil.IsGlob(glob) {
-		// glob represent plain object
-		handler.FileNames = []string{globPattern}
-		err := handler.DownloadAll(ctx)
-		if err != nil {
-			return nil, err
-		}
-		return handler.LocalPaths, nil
+	tempDir, err := os.MkdirTemp(os.TempDir(), "blob*ingestion")
+	if err != nil {
+		return nil, err
 	}
 
-	listOptions := &blob.ListOptions{BeforeList: func(as func(interface{}) bool) error {
-		// Access storage.Query via q here.
-		var q *storage.Query
-		if as(&q) {
-			// we only need name and size, adding only required attributes to reduce data fetched
-			_ = q.SetAttrSelection([]string{"Name", "Size"})
-		}
-		return nil
-	}}
+	return &blobIterator{
+		ctx:     ctx,
+		bucket:  bucket,
+		objects: items,
+		tempDir: tempDir,
+	}, nil
+}
 
-	if prefix != "." {
-		listOptions.Prefix = prefix
+func (iter *blobIterator) Close() error {
+	fileutil.ForceRemoveFiles(iter.localFiles)
+	// remove temp dir since recursive paths created have to be removed as well
+	os.RemoveAll(iter.tempDir)
+	return iter.bucket.Close()
+}
+
+func (iter *blobIterator) HasNext() bool {
+	return iter.index < len(iter.objects)
+}
+
+// NextBatch downloads next n files and copies to local directory
+// Callers responsibility to delete files once done
+// Thread unsafe
+func (iter *blobIterator) NextBatch(n int) ([]string, error) {
+	if !iter.HasNext() {
+		return nil, io.EOF
 	}
 
-	var size, fetched int64
-	var matchCount int
-	var fileNames []string
+	// delete files created in last iteration
+	fileutil.ForceRemoveFiles(iter.localFiles)
+	start := iter.index
+	end := iter.index + n
+	if end > len(iter.objects) {
+		end = len(iter.objects)
+	}
+	iter.index = end
+
+	// this is not necessary on every iteration
+	// but there may be cases where n in first batch is different from n in next batch
+	// to keep things easy creating a new slice every time
+	iter.localFiles = make([]string, n)
+	g, grpCtx := errgroup.WithContext(iter.ctx)
+	g.SetLimit(_concurrentBlobDownloadLimit)
+	for i, obj := range iter.objects[start:end] {
+		obj := obj
+		index := start + i // with repect to object slice
+		g.Go(func() error {
+			// need to create file by maintaining same dir path as in glob for hivepartition support
+			dir := filepath.Join(iter.tempDir, filepath.Dir(obj.obj.Key))
+			// filename
+			objName := filepath.Base(obj.obj.Key)
+			file, err := fileutil.OpenTempFileInDir(dir, objName)
```

Comment:
That doesn't work since we need to create file by maintaining same dir path as in glob for hivepartition support and openFile doesn't create the required directories. 
Example : Imagine having object as s3://<bucket>/path/year=2020/obj.parquet
we want to create file with the same path in tempDir. 
Example: /tmp/blob123ingestion/year=2020/obj.parquet

Reply:
That makes sense, we need the `MkdirAll` as well. But since `tempDir` is initialized with `MkdirTemp`, it will already be a unique name. So there's no need for `OpenTempFileInDir`, which will also add random chars to the filename (in your example, it will be `/tmp/blob123ingestion/year=2020/obj-RANDOMRANDOM.parquet`.

So something like this would be sufficient, right:

```go
// iter.tempDir is previously initialized with MkdirTemp
filename := filepath.Join(iter.tempDir, obj.obj.Key)
os.MkdirAll(filepath.Dir(filename))
file := os.OpenFile(filename)
```

Reply:
Yeah makes sense.